lib/io.ml

```diff
@@ -40,8 +40,12 @@ let write fd buf =
 let connect host port =
   let sockaddr = Unix.ADDR_INET (host, port) in
   let fd = Unix.socket Unix.PF_INET Unix.SOCK_STREAM 0 in
-  Unix.connect fd sockaddr;
-  fd
+  try
+    Unix.connect fd sockaddr;
+    fd
+  with e ->
+    Unix.close fd;
+    raise e
```

Comment:
Now that I think about it, in `xenopsd` we already use `xapi-stdext-pervasives`, so you should probably use `Xapi_stdext_pervasives.Pervasiveext.finally`: `finally (fun () -> Unix.connect fd sockaddr; fd) (fun () -> Unix.close fd)`

Reply:
No, this is not the case that we could apply `finally` here. If no error happens, we should NOT close the `fd`.

Reply:
Oh geez, you are right. I was not thinking straight!